### DISCUSSION TITLE
Orphaned Containers??

### DISCUSSION THREAD

**Discussion by rfpludwick** [concourse/concourse#8553] [Help & Support] [answered]
I keep hitting limits like `find or create container on worker concourse-worker-1: new container: checking container capacity: max containers reached` after just a few builds running. In debugging, it seems like Concourse's internal worker container tracking doesn't match up, at least according to the output below. What am I missing here? `concourse-worker-1` reports 250 containers, but the `fly containers` output shows nowhere near that many.

```shell
❯ fly -t cluster workers
name                containers  platform  tags  team  state    version  age
concourse-worker-0  41          linux     none  none  running  2.4      13h3m
concourse-worker-1  250         linux     none  none  running  2.4      13h4m
concourse-worker-2  33          linux     none  none  running  2.4      13h4m

❯ fly -t cluster containers
handle                                worker              pipeline                          job                                            build #  build id  type   name                         attempt
00d8018e-462c-4842-6e8f-18083911edbd  concourse-worker-1  docker-concourse-ci-worker-build  build                                          23       5577      task   build-image                  n/a
2a69081e-e333-4e01-6ad5-d19434180961  concourse-worker-0  docker-concourse-ci-worker-build  build                                          23       5577      check  image                        n/a
3301d0f5-0a49-41c8-7946-8e744cca5691  concourse-worker-0  docker-concourse-ci-worker-build  build                                          24       5617      get    image                        n/a
36de1300-3305-49d6-4d52-8c2b847c4c5b  concourse-worker-1  docker-concourse-ci-worker-build  build                                          23       5577      get    repo                         n/a
42c325ed-da51-49e8-711d-bd20a4987b61  concourse-worker-0  docker-concourse-ci-worker-build  build                                          22.1     5534      get    image                        n/a
457acbc7-b857-41b1-6ce7-9d1a4fe1060a  concourse-worker-2  docker-concourse-ci-worker-build  build                                          22.1     5534      task   build-image                  n/a
5f61d289-ad81-40bd-5c5c-5b6d4b9f1639  concourse-worker-1  docker-concourse-ci-worker-build  build                                          24       5617      task   build-image                  n/a
639ede0a-6eb1-4d28-5592-59a4c1ffde6c  concourse-worker-0  docker-concourse-ci-worker-build  build                                          22.1     5534      task   get-datetime                 n/a
65ce8dc8-c7f6-4a71-48ea-c598a7298a7d  concourse-worker-0  all-branches                      set-pipeline-docker-concourse-ci-worker-build  12       5615      get    repo-concourse-ci-pipelines  n/a
7cfa58bb-6247-4d36-4619-9b5f7a65b961  concourse-worker-1  docker-concourse-ci-worker-build  build                                          23       5577      task   get-datetime                 n/a
7e9a9da0-e156-468d-7129-14a834dc9dff  concourse-worker-1  docker-concourse-ci-worker-build  build                                          22.1     5534      check  image                        n/a
7f56548c-8fbe-4a5c-75c4-b48d0e6f77fe  concourse-worker-2  all-branches                      none                                           check    none      check  repo-concourse-ci-pipelines  n/a
814333f9-d7b4-4208-4a80-2e60af8d3ed5  concourse-worker-2  docker-concourse-ci-worker-build  build                                          23       5577      get    image                        n/a
85c82469-c41a-4d55-672a-3a39d88956e9  concourse-worker-2  docker-concourse-ci-worker-build  build                                          24       5617      task   get-datetime                 n/a
868fdfdd-89b0-446c-7fa1-fbe4ef562241  concourse-worker-0  docker-concourse-ci-worker-build  build                                          24       5617      check  image                        n/a
888ab39b-6d71-475c-548a-5d42de1d1b79  concourse-worker-1  docker-concourse-ci-worker-build  build                                          22.1     5534      get    image                        n/a
93e94e3e-53d1-484a-619e-1e6a1d055bd4  concourse-worker-2  all-branches                      set-pipeline-docker-concourse-ci-worker-build  11       5570      get    repo-concourse-ci-pipelines  n/a
9aee1201-d2b1-44cf-52f2-c8e3a534fd8d  concourse-worker-0  docker-concourse-ci-worker-build  build                                          22.1     5534      check  image                        n/a
b84d0f80-c6d6-4c9c-402e-2e615fddc735  concourse-worker-2  docker-concourse-ci-worker-build  build                                          23       5577      check  image                        n/a
dc9a6ada-317b-41b3-4566-185f2ed55143  concourse-worker-2  docker-concourse-ci-worker-build  none                                           check    none      check  repo                         n/a
e75b68fe-ad81-4952-5ff7-4321b853f9b2  concourse-worker-2  docker-concourse-ci-worker-build  build                                          24       5617      check  image                        n/a
fdb7798e-70bb-4799-4d55-3697d76b9f57  concourse-worker-2  docker-concourse-ci-worker-build  build                                          22.1     5534      get    repo                         n/a
```

**Comment by marco-m** [comment]
Command `fly containers` reports the _active_ containers (containers you can fly intercept), while `fly workers` reports all the containers. All containers will be, sooner or later, garbage collected.

It is a good idea to read https://concourse-ci.org/container-placement.html and choose one (or multiple) placement strategy that match (more or less) the usage pattern of your deployment.

When the number of containers on a worker goes above circa 200--220, often things starts going sour.

It is also fundamental to setup metrics collection and monitoring, both to get an intuitive feeling of how the system behaves and to take informed decisions when things go bad...

**Reply by rfpludwick** [reply]
Is there any way to force garbage collection to be more aggressive?

**Reply by rfpludwick** [reply]
I ask on the GC front because I've got this set in my Helm chart's `values.yaml` and it doesn't seem to be having a positive imact:

```yaml
concourse:
  web:
    gc:
      hijackGracePeriod: 5m
      failedGracePeriod: 5m
      checkRecyclePeriod: 5m
      varSourceRecyclePeriod: 5m
```

**Reply by rfpludwick** [reply]
@marco-m Do you happen to have any insight here? I've got three workers set to 200 container max each and they hit that limit within about 3-4 hours and never drop down below. My GC intervals are all set to 5 minutes, which to me would seem to indicate that they ought to be getting cleaned up pretty quickly, but it doesn't seem to happen.

**Comment by rfpludwick** [comment, marked as answer]
Finally found the problem here - cgroups v2 is not well supported yet, so I had to downgrade to cgroup v1.

**Reply by mrinc** [reply]
We're 3y on and I am now having this issue ... were you able to go back to cgroups v2 or still stuck on v1?

**Reply by rfpludwick** [reply]
I haven't tried a newer version of Concourse; there are no readily available ARM64 images, which is the hardware I'm using. If the newer versions better support cgroups v2, that'd be great since it's now in maintenance support in Kubernetes.

**Reply by taylorsilva** [reply]
Do you have more details? What made you say
> cgroups v2 is not well supported

If you have more details, like error logs from the workers, that would be helpful as I haven't personally seen any issues with running Concourse on cgroup v2 workers where cgroups v1 is completely disabled (e.g. Fedora 39+).

---

Comment more related to your original post: it sounds like the worker may be having issues garbage collecting containers. I'd first look for errors on the worker related to not being able to delete containers.

Also, it's worth noting that Concourse keeps track of containers in its database. It also asks workers for the list of containers they currently know about. Any containers that workers report back to Concourse that Concourse doesn't have in its db are then told to be deleted. Logic looks like this:

> Web node: I know about containers A and B. Hey Worker, what containers do you have?
Worker: I have containers A, B, C, E
Web node: I don't have any info about containers C and E. Worker, delete containers C and E.
Worker: Okay! Deleting containers C and E

**Reply by rfpludwick** [reply]
@taylorsilva I don't have any logs from the workers; this issue cropped up for me over 2 years ago.

**Reply by rfpludwick** [reply]
@taylorsilva I can confirm now that I've upgraded to the latest Concourse ARM version that this is no longer a problem with v2.

**Reply by taylorsilva** [reply]
Glad to hear! Thanks for letting us know :)
